Empty Pockets / Negative Search Results

qoala opened the pull request:
* The 'steal' action is also available if the attempt would reveal negative information ("the alarm 3+ guard wasn't carrying anything", "a previously searched guard doesn't have anything more now that an anarchy-5 agent is here"), allowing the "Searched" and "Expertly Searched" flags to be set.
* When doing so, a "NOTHING FOUND" popup appears to clarify what happened.
* NEED A DOLLAR is superceded by this, so no longer appears as a campaign option. The implementation code remains to avoid breaking existing saves with the option enabled.

If we instead want to preserve NEED A DOLLAR as an option, could add the additional [Sim Constructor](https://steamcommunity.com/workshop/filedetails/?id=580661011) argument, specifying `{default=false}`. If Sim Constructor isn't installed, lua should ignore the extra function parameter and the option would default to true like normal.

qoala committed:
Copy vanilla simquery.canLoot

qoala committed:
modify canLoot to allow searching for negative results

qoala committed:
Replace need_a_dollar with empty_pockets in generation options

qoala committed:
Add a popup when search finds negative information

qoala committed:
Cleanup need_a_dollar option code

The implementation remains to avoid breaking existing games, but can no
longer be enabled.

qoala committed:
Clear 'need_a_dollar' if present on a new campaign

Customizable Generation Presets can preserve options from previous
verions, which can throw off "retain behavior for existing saves".
Luckily CGP requires SimConstructor, and SimConstructor indicates when a
new campaign is being started.

qoala commented:
Further testing revealed an issue with [Customizable Generation Presets](https://steamcommunity.com/sharedfiles/filedetails/?id=1631664146):
Generation Presets snapshots all currently available options and allows reapplying them on future runs. This PR was removing `need_a_dollar` from the UI but respecting it if set, to avoid breaking existing saves. Generation Presets was adding `need_a_dollar` on new campaigns.

The fix was to use a feature of Sim Constructor (required for Generation Presets) to detect when a new campaign is started. The argument is nil without Sim Constructor installed, so this fix won't apply to games without those mods.

osheroff commented:
sorry, I read through your explanation a couple of times without quite getting what this code does.  can you clarify?

qoala commented:
I'm not sure which piece you're asking to clarify, so hopefully it's answered below:
* `canLoot` changes:
  * `simquery.canLoot` is exclusively used to determine whether or not the "Steal" action appears in the UI. The generic "Steal" UI action applies to pickpocketing KOed and killed guards/drones, but not safes. Safes and similar items have a separate code path. "Steal" triggers a "search" action that marks searched/expertly-searched and opens a UI for further inventory actions.
  * Extracted the "does target have something to steal" check into a submethod, to which was added overrides:
    * the target does not yet have "searched" -> Attempting to steal will usefully mark the target as "searched"
    * the agent has anarchy 5 and the target does not yet have "expertly searched" -> Attempting to steal will usefully mark the target as "expertly searched"
  * This provides the benefits of "need a dollar" without the extra credit. It also provides similar benefits for expertly-searching guards, which wasn't provided by "need a dollar".
* `lootPanel` changes:
  * This is the UI element for stealing from guards/corpses and safes. Its `:refresh()` automatically closes the panel when the target inventory is empty, and does not normally appear when the target inventory is empty.
  * Added a flag to detect the first UI refresh upon spawning the panel. This prevents the change from accidentally firing after the user steals everything from the target.
  * Upon the first UI refresh of the panel, if there is nothing to steal:
    * close the panel.
    * show a "nothing to steal" pop-up instead, to inform the user that "Steal" failed and why.
* `modinit` changes in the last commit:
  * The SimConstructor mod modifies the `modApi` with a number of features to help modders, among them a `params` table that can be populated during the first `load()` of a new campaign (and is `nil` when loading an existing campaign) and is then available during the game via `sim:getParams().difficultyOptions`. This serves a similar purpose to `monkey_patch.lua` in controlling whether or not a change should be available based on the campaign options. UI Tweaks currently doesn't require Sim Constructor, and I'm not changing that.
  * The Generation Presets mod modifies the "New Campaign" screen to save and restore the state of all mod campaign options. This can preserve the `need_a_dollar` setting even though it's no longer available in UI Tweaks' `init()`.
  * In `load( modApi, options, params )`, `params` is only truthy iff Sim Constructor is installed and this is a new campaign
    * Without Sim Constructor (and by dependency, without Generation Presets): lua passes `nil` for arguments the caller didn't provide. If `need_a_dollar` is present, it's legitimately part of an existing save and should be respected.
    * With Sim Constructor, loading an existing campaign: Sim Constructor passes `params = nil`. If `need_a_dollar` is present, it's legitimately part of an existing save and should be respected.
    * With Sim Constructor, loading a _new_ campaign: Sim Constructor passes a table for `params` (truthy in lua). This code's `init()` didn't define `need_a_dollar`, so if it's present, then Generation Presets preserved it into the new campaign. I only intend to preserve its behavior in existing saves, so in this case, the option is cleared.

---

Slight quirks noticed in my twitch stream this past weekend (I've been playing with the changes to try and shake out any potential bugs):

1. The game doesn't preserve searched/expertly-searched when a guard dies and becomes a "corpse" unit. This PR's code correctly invites the user to attempt stealing from the corpse in order to mark it "searched", even though the user likely knows that the guard already had nothing left to steal.
    * Ideally, we'd modify the corpse-generation code to preserve the searched traits, but that code is buried in a function with legitimate reasons for other mods to modify it (`simunit:killUnit`). Fixing that would be more involved.

2. A modded enemy device (shock mines are floor traps from Worldgen Extended) technically passes the criteria for the vanilla `canLoot` other than never carrying items/credits/pwr (it can be destroyed with weapons, so qualifies as a "target", and unlike cameras/turrets is flagged as "enemy" not "neutral"). It's somewhat erroneous that this PR's code invites the user to mark it "searched", because the user likely doesn't expect that to be meaningful.
    * Adding a `(target:isAgent() or target:isCorpse())` clause on this PR's checks would constrain searching to enemies that the player expects "Steal" is applicable (guards, drones, and corpses).

I'll fix the 2nd issue on another evening when I have more time.

osheroff commented:
ok.  seems fine enoug.  fix the conflicts and I'll merge (and figure out how to release a new version)

BTW, I'm playing invisible again (if only to ween myself off my horrific slay the spire addiction, aka video game methodone), and the situation came up where an agent really wanted to path into a watched spot, even though there was a safe (and equal route).  Installed your latest PR and worked like a charm!

qoala committed:
Merge remote-tracking branch 'upstream/master' into emptypockets

qoala committed:
Only show empty-pockets prompts for guards/drones and corpses

For compatibility with Worldgen Extended, where shock traps qualify as
"enemy targets" but the player does not expect "Steal" to be applicable.

qoala commented:
Fixed merge conflict, and restricted the `canLoot` condition so that it doesn't show for Worldgen Extended's shock mines.

(StS is definitely another good game to get sucked into)

In Steam, there's a separate "Invisible, Inc. Mod Uploader" entry. (You may need to enable "tools" in the "games" filter to see it) Select the mod to update, and point the tool at the `out/` folder.

qoala commented:
Any updates?
